### PR TITLE
feature: support tool calling from compiler and gateway

### PR DESCRIPTION
**Pull request by geclos**

### CODE REVIEW

File: packages/compiler/src/compiler/base/nodes/tags/message.ts [hunk 1]
```diff
@@ -120,7 +120,6 @@ function buildMessage<R extends MessageRole>(
 
     return {
       role,
-      id: String(attributes.id),
```
Comment (geclos) [inline packages/compiler/src/compiler/base/nodes/tags/message.ts:123]: not sure what this id is tbh don't see it anywhere in ai sdk documentation

File: apps/gateway/src/common/messageSchema.ts [hunk 2]
```diff
@@ -32,13 +32,22 @@ export const messageSchema = z
...
+            type: z.literal(ContentType.toolCall),
+            toolCallId: z.string(),
+            toolName: z.string(),
+            args: z.record(z.any()),
+          }),
+        ),
+      ),
       toolCalls: z.array(
         z.object({
           id: z.string(),
```
Comment (csansoon) [inline apps/gateway/src/common/messageSchema.ts:53]: `toolCalls` is not part of the `content`, is it? An assistant message could contain both a set of `toolCalls` and a `content` string.

File: packages/core/src/helpers.ts [hunk 3]
```diff
@@ -9,6 +9,8 @@ export function objectToString(
   message = DEFAULT_OBJECT_TO_STRING_MESSAGE,
 ) {
   try {
+    if (!object) return ''
```
Comment (geclos) [inline packages/core/src/helpers.ts:12]: we were stryingifying stuff like null into 'null'

File: apps/gateway/src/common/messageSchema.ts (hunk 2 above)
Comment (geclos) [reply apps/gateway/src/common/messageSchema.ts:53]: nein, it can also contain an array of tool calls within the content in fact

Comment (csansoon) [reply apps/gateway/src/common/messageSchema.ts:53]: At least according to openai schema. Is vercel's ai different?

File: apps/gateway/src/common/messageSchema.ts [hunk 4]
```diff
@@ -51,7 +60,14 @@ export const messageSchema = z
...
-      id: z.string(),
+      content: z.array(
+        z.object({
+          type: z.literal(ContentType.toolResult),
+          toolCallId: z.string(),
+          toolName: z.string(),
+          result: z.string(),
+          isError: z.boolean().optional(),
+        }),
+      ),
```
Comment (csansoon) [inline apps/gateway/src/common/messageSchema.ts:71]: A `tool` message should only contain an `id` and `content`, at least according to the openai schema. Why do we need the rest?

File: packages/compiler/src/compiler/base/nodes/tags/message.ts [hunk 5]
```diff
@@ -120,7 +120,6 @@ function buildMessage<R extends MessageRole>(
...
       content,
     } as ToolMessage
   }
```
Comment (csansoon) [inline packages/compiler/src/compiler/base/nodes/tags/message.ts:125]: Why? A `tool` message requires an `id`, which identifies the `tool_call` that requested it

Comment (geclos) [reply packages/compiler/src/compiler/base/nodes/tags/message.ts:125]: the schema is just wrong this is the valid schemas both for tool result and tool calling requested https://sdk.vercel.ai/docs/reference/ai-sdk-core/stream-text#messages-role not sure where this id thing came from

File: apps/gateway/src/common/messageSchema.ts (hunk 4 above)
Comment (geclos) [reply apps/gateway/src/common/messageSchema.ts:71]: not true according to ai sdk documentation (and it wasn't working either)

File: apps/gateway/src/common/messageSchema.ts (hunk 2 above)
Comment (geclos) [reply apps/gateway/src/common/messageSchema.ts:53]: yes it's different

File: packages/compiler/src/compiler/base/nodes/tags/message.ts (hunk 1 above)
Comment (csansoon) [reply packages/compiler/src/compiler/base/nodes/tags/message.ts:123]: It is required in the openai schema. I don't know how vercel would manage this if it is not included when openai requires it...

File: packages/compiler/src/compiler/base/nodes/tags/message.ts (hunk 5 above)
Comment (csansoon) [reply packages/compiler/src/compiler/base/nodes/tags/message.ts:125]: This differs from openai's schema, which is what the compiler is based on
https://platform.openai.com/docs/api-reference/chat/create#chat-create-messages

Do we want to change the compiler's schema to fit vercel's instead?

File: packages/compiler/src/compiler/base/nodes/tags/message.ts (hunk 1 above)
Comment (geclos) [reply packages/compiler/src/compiler/base/nodes/tags/message.ts:123]: talked offline

File: packages/compiler/src/compiler/base/nodes/tags/message.ts (hunk 5 above)
Comment (geclos) [reply packages/compiler/src/compiler/base/nodes/tags/message.ts:125]: talked offline

**Pull request merged**